dali/pipeline/data/types.h

```diff
@@ -465,42 +467,54 @@ class DLL_PUBLIC TypeTable {
 
   template <typename T>
   DALIDataType RegisterType(DALIDataType dtype) {
-    std::lock_guard<spinlock> guard(lock_);
-    // Check the map for this types id
-    auto id_it = type_map_.find(typeid(T));
-
-    if (id_it == type_map_.end()) {
-      type_map_[typeid(T)] = dtype;
-      TypeInfo t;
-      t.SetType<T>(dtype);
-      type_info_map_[dtype] = t;
+    static DALIDataType id = [dtype, this]() {
+      std::lock_guard guard(insert_lock_);
+      size_t idx = dtype - DALI_NO_TYPE;
```

Comment:
What's the point of `- DALI_NO_TYPE` part (which is -1)?

Reply:
The idea is that the "type id" is in fact just an _index_ - so I just subtract the smallest possible id and get 0-based index into the type array.

Reply:
Oh, ok! I though so, but somehow didn't realize we may want to map the DALI_NO_TYPE as well.